`General`: Allow admins to filter user management table

Dominik-Weinzierl opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Prepare #5061 and provides a filterable user management view

### Description
<!-- Describe your changes in detail -->
In order to still provide the possibilities of pages, i integrated the filter options into the page and filter the users on the database level. This allows us to still use the benefits of pages.

We provide several filters:

- **Role**: Filter the users according to their permissions/role _(stored in local storage)_
- **Origin**: Filters the users based on their administrative location (internal or external) _(stored in local storage)_
- **Status**: Filter the users based on their account status (activated or deactivated) _(stored in local storage)_
- **Course**: Filters the users according to their course affiliation

The first time we visit the page, we have to load all the courses (which takes some time). After that, the filtering should be really fast.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Admin

1. Log in to Artemis
2. Visit the user management page
3. Add/Remove users, apply filters and so on
4. Verify that everything works

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
| Class/File | Branch | Line |
|------------|-------:|-----:|
| user-management.component.ts | 73.68% | 83.81% |
| UserResource | 89% | 75% |

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

![image](https://user-images.githubusercontent.com/60005702/172572876-f65b39db-e1ef-467e-a97c-0483c17f29dc.png)
![image](https://user-images.githubusercontent.com/60005702/172572943-99fd3cf6-c16e-48dd-acb9-473d5745d14c.png)
![image](https://user-images.githubusercontent.com/60005702/172572991-11406c6b-b8c4-4955-ae94-4cb23f372294.png)

krusche commented:
I think this functionality is too specific for a special button. If we add such buttons, we will further decrease the usability of the system.

One alternative and better approach could be to offer a filter functionality, then an admin could still delete the filtered users.

The mentioned sub group of users would be a combination of multiple filters.

We could then offer bulk actions as we do for programming exercises to support the case of deleting all filtered users

dfuchss commented:
I guess filters and bulk actions would be good.

Dominik-Weinzierl commented:
@krusche @dfuchss 
![image](https://user-images.githubusercontent.com/60005702/171406272-b69ca179-530c-4798-b1ac-f4bf12505ecb.png)

- **Role**: Roles/Authorities of the user (e.g. Admin, Editor, ..)
- **Origin**: Internal or external user (or both)
- **Status**: Activated/Deactivated users/accounts 

Does that fit your/our needs? :thinking:

dfuchss commented:
So maybe add a course filter where you can simply check for selected courses (checkboxes, all, none)

b-fein commented:
I suppose the filter for internal/external users is relevant regardless if only one or both can be enabled, as it might be possible that a specific Artemis installation switched from one authentication method to another at some point. Then there maybe exist old user accounts of the no longer used type. An admin might want to filter for those.

Dominik-Weinzierl commented:
@dfuchss any ideas where to add the all/none checkbox? :D otherwise you need to deselect all courses manually 
![image](https://user-images.githubusercontent.com/60005702/171487715-dc35ced8-6d1f-4378-afa2-194bd43da528.png)

dfuchss commented:
I think a none checkbox might be needed since you may want an option do not filter by courses 

So you need the ability to select "No Course" and the ability to not use the filter.

dfuchss commented:
Yes that would work. The semantic of the filter for the courses would be "Course X or Course Y" if >0 courses are selected (? or would that be an "and"; I'm not sure here). 
If 0 courses are selected the semantic could be that you're looking for people without any course.

krusche commented:
Just looking at the screenshot, I do not really know what origin and status means. Also the filtering part becomes quite large. I would suggest we hide it in one filter button. When pressing on it, all options are shown in a popup/tooltip windows, the apply button is then also part of the popup/tooltip window. An Alternative could be that the filter area can be collapsed.

Can you think about this? And please also post a screenshot of the whole page.

krusche commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -147,9 +149,44 @@ public interface UserRepository extends JpaRepository<User, Long> {
-    @Query("select user from User user where user.login like %:#{#searchTerm}% or user.email like %:#{#searchTerm}% "
-            + "or user.lastName like %:#{#searchTerm}% or user.firstName like %:#{#searchTerm}%")
-    Page<User> searchByLoginOrNameWithGroups(@Param("searchTerm") String searchTerm, Pageable pageable);
+    @Query("""
```
Maybe a solution like pointed out here would be better: https://stackoverflow.com/questions/58240024/search-by-many-optional-parameters-in-spring-data-jpa-repository

We don't use something like this at the moment, but it's surely a good opportunity to try out. We have other places with similar problems 

@canberkanar maybe this is also an option for the filters with sub queries for Metis?

Dominik-Weinzierl replied:
Thanks for the suggestion :+1: I will try that out.

Dominik-Weinzierl replied:
Addressed in c9e1373b1a59b58037e90512ea3cfed226a376d0 - sadly we have still subqueries, but i guess the code is now more readable and extendable (modular specifications) :+1:

krusche replied:
sub queries are not allowed! please avoid them

Dominik-Weinzierl replied:
Yes, this is an outdated comment - they should be avoided if possible :+1:

TheTrueVester reviewed (approved):
Tested on ts1. Everything worked as expected. 
As a recommendation: add a way to ignore the Course filter to be able to see all Users at the same time.
![image](https://user-images.githubusercontent.com/47668936/172072172-9e322def-8991-4438-b439-19c601ecf5dc.png)

babyygemperor reviewed (dismissed):
Tested on ts2, works as described. 
Just one small note, it does look a bit odd, that the filter button only shows up once the list has loaded. The change in UI looks a bit jarring
<img width="2560" alt="image" src="https://user-images.githubusercontent.com/25747019/172077182-f80359f1-7fcb-4078-adaf-4b3e8aa04587.png">

Dominik-Weinzierl commented:
@TheTrueVester thanks for the suggestion - i will check that :+1: 
@babyygemperor since the table is not created until the data is loaded, and the filter and the search bar are both part of the table, they are not displayed - but i will keep that in mind. Thank you!

Dominik-Weinzierl commented:
@TheTrueVester @babyygemperor i refactored the complete filter logic. Now the filter works as filter and users must match **ALL** selected values and not only one of the selected values :+1: (which is the normal filter logic)

nairaarnez reviewed (commented):
I couldn't find the users page, could you provide the path to it please? It is not this one, right?
![image](https://user-images.githubusercontent.com/104493356/172164572-399081f6-0e59-45b3-a06b-b0344937bcf4.png)

Dominik-Weinzierl commented:
@nairaarnez you need to be admin and the link is `/admin/user-management`

pantageepapa reviewed (approved):
Tested on ts2 and everything worked as expected, which means the elements have to fulfill all selected criteria to be shown. 

However, my suggestion would be to apply OR logic inside the category and AND logic between categories. It was quite unusual for me that clicking on "All" of the category "Role" filtered out every user.

Dominik-Weinzierl commented:
@pantageepapa i guess it depends on the use case. Do i want to get all users which are part of at least one of the selected courses or do i want to get the users which are part of all selected courses :thinking:

mallgaier reviewed (dismissed):
Tested on TS1, works as expected.
Great job 👍

dfuchss commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -73,7 +75,7 @@ public interface UserRepository extends JpaRepository<User, Long> {
     @Query("select count(*) from User user where :#{#groupName} member of user.groups")
     Long countByGroupsIsContaining(@Param("groupName") String groupName);
 
-    @Query("select user from User user where lower(user.email) = lower(:#{#searchInput}) or lower(user.login) = lower(:#{#searchInput})")
+    @Query("select user from User user where lower(user.email) = lower(:#{#searchInput}) or lower(user.login) = lower" + "(:#{#searchInput})")
```
Why this change :) ?

Dominik-Weinzierl replied:
I guess that was not intended :D

dfuchss commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -95,7 +97,7 @@ public interface UserRepository extends JpaRepository<User, Long> {
      */
     @EntityGraph(type = LOAD, attributePaths = { "groups" })
     @Query("select user from User user where :#{#groupName} member of user.groups and "
-            + "(user.login like :#{#loginOrName}% or concat_ws(' ', user.firstName, user.lastName) like %:#{#loginOrName}%)")
+            + "(user.login like :#{#loginOrName}% or concat_ws(' ', user.firstName, user.lastName) like " + "%:#{#loginOrName}%)")
```
Is this for better readability ?

dfuchss commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -174,21 +172,221 @@ public interface UserRepository extends JpaRepository<User, Long> {
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getCourseSpecification(Set<Long> courseIds) {
+        if (courseIds.size() == 1 && courseIds.contains(-1L)) {
```
I'm not sure here, but I would suggest to create a constant for "-1L" like `COURSE_ID_FOR_EMPTY_COURSES` that makes the purpose of the -1 clear

Dominik-Weinzierl replied:
Good point :D Thank you!

dfuchss reviewed (commented):
Code lgtm.
I made some minor remarks :)

I'll test it the next days, then I'll approve it.

bassner commented on src/main/webapp/app/admin/user-management/user-management.component.ts:
```diff
@@ -103,6 +190,187 @@ export class UserManagementComponent implements OnInit, OnDestroy {
+        this.filters.originFilter = this.initFilter(UserStorageKey.ORIGIN, OriginFilter) as Set<OriginFilter>;
+        this.filters.statusFilter = this.initFilter(UserStorageKey.STATUS, StatusFilter) as Set<StatusFilter>;
+
+        let key = this.localStorage.retrieve(UserStorageKey.NO_COURSE);
+        this.filters.noCourse = key ? (key as boolean) : false;
+
+        key = this.localStorage.retrieve(UserStorageKey.NO_AUTHORITY);
+        this.filters.noAuthority = key ? (key as boolean) : false;
```
```suggestion
        this.filters.noCourse = !!this.localStorage.retrieve(UserStorageKey.NO_COURSE);
        this.filters.noAuthority = !!this.localStorage.retrieve(UserStorageKey.NO_AUTHORITY);
```

bassner commented on src/main/webapp/app/admin/user-management/user-management.component.ts:
```diff
@@ -103,6 +190,187 @@ export class UserManagementComponent implements OnInit, OnDestroy {
+     * @param key of the filter in the local storage
+     * @param type of filter
+     */
+    initFilter(key: UserStorageKey, type: any) {
+        const temp = this.localStorage.retrieve(key);
+        const tempInStorage =
+            temp !== undefined && temp !== null
+                ? temp
+                      .split(',')
+                      .map((filter: string) => type[filter])
+                      .filter(Boolean)
+                : new Set();
+        return new Set(tempInStorage);
+    }
```
I suggest this to simplify the inner logic a bit, prevent the use of ``any``, and give it a proper return type. There could be better ways to do it than this, but we should avoid using ``any`` everywhere if possible

```suggestion
    initFilter<E>(key: UserStorageKey, type: { [key: string]: string }): Set<E> {
        const temp = this.localStorage.retrieve(key);
        const tempInStorage =
            temp
                ?.split(',')
                .map((filter: string) => type[filter])
                .filter(Boolean) ?? [];
        return new Set(tempInStorage);
    }
```

You can replace the places a few lines above where you use the method like and thereby get rid of the cast:
```js
        this.filters.authorityFilter = this.initFilter<AuthorityFilter>(UserStorageKey.AUTHORITY, AuthorityFilter);
        this.filters.originFilter = this.initFilter<OriginFilter>(UserStorageKey.ORIGIN, OriginFilter);
        this.filters.statusFilter = this.initFilter<StatusFilter>(UserStorageKey.STATUS, StatusFilter);
```

In general, there's a lot of ``any`` below. You could try similar solutions there to get rid of them, e.g. by containing a combined type of your filter enums.

Dominik-Weinzierl replied:
Thanks for your suggestions: Solved in cb951a1a5f3c417e2126fb154ff154eab965a1d3 :D

bassner reviewed (changes requested):
Reviewed the client code and left two remarks / suggestions below. :)

TheTrueVester reviewed (approved):
Tested on ts2. Everything worked as expected. 
I did however notice that it's possible to filter for accounts that are activated and deactivated at the same time (leading to no results). This is probably something that can never happen under normal circumstances. 
![image](https://user-images.githubusercontent.com/47668936/172366666-e212708d-c258-4961-8bf3-89e1eab36eb1.png)

Dominik-Weinzierl commented:
@TheTrueVester I updated the buttons to radio buttons - now it is not possible anymore to select both states :+1: In addition, i also added a `None` button which represents no filter for this category.

rhubinger reviewed (approved):
Tested on TS2. Works as described.

krusche commented:
Does the Filter button show how many filters are currently active? I can not see this in the screenshots

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -133,9 +136,9 @@ public interface UserRepository extends JpaRepository<User, Long> {
      *
      * @param page        Pageable related info (e.g. for page size)
      * @param loginOrName Either a login (e.g. ga12abc) or name (e.g. Max Mustermann) by which to search
-     * @return            list of found users that match the search criteria
+     * @return list of found users that match the search criteria
```
I think this formatting was correct, see e.g. this change: https://github.com/ls1intum/Artemis/pull/5163/files#r882969773

Dominik-Weinzierl replied:
Reverted - thanks to my autoformater :P

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -174,21 +173,221 @@ public interface UserRepository extends JpaRepository<User, Long> {
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getAuthoritySpecification(Set<String> authorities) {
+        if (authorities.contains("NO_AUTHORITY")) {
```
Can you use some constants instead of strings here and below?

Dominik-Weinzierl replied:
Sure :D

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -174,21 +173,221 @@ public interface UserRepository extends JpaRepository<User, Long> {
+        Set<String> authorities = userSearch.getAuthorities();
+
+        // Internal or external users or both
+        final var internal = userSearch.getOrigins().contains("INTERNAL");
```
same as above.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/web/rest/dto/UserPageableSearchDTO.java:
```diff
@@ -0,0 +1,61 @@
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class UserPageableSearchDTO<T> extends PageableSearchDTO<T> {
```
I don't know what the exact purpose of the generic here is for, but from the looks of it you always use it with Strings in the rest of the code. Will it be always a String? Or do you need this parameter to be generic?
```suggestion
public class UserPageableSearchDTO extends PageableSearchDTO<String> {
```

Dominik-Weinzierl replied:
Good point - i guess we can remove the generic :D

Dominik-Weinzierl commented:
@krusche @Strohgelaender The filter button is adapted and the suggestions implemented now :+1: (and the screenshots updated).

dfuchss commented:
@Dominik-Weinzierl do you have also implemented the possibility to execute some bulk actions (because I cannot find this).
If not, I think it does not close #5061 , because I cannot click 800 times in the table :)

Dominik-Weinzierl commented:
@dfuchss yes you are right - this PR prepares the bulk deletion but does not implement it :+1: i adapted the description accordingly :+1:

dfuchss reviewed (dismissed):
Tested on our test systems.

dfuchss commented:
> @dfuchss yes you are right - this PR prepares the bulk deletion but does not implement it 👍 i adapted the description accordingly 👍

Perfect. Approved now :)

bassner commented:
Are the CodeQL and Codacy suggestions valid?

rhubinger reviewed (dismissed):
Tested on TS2. Works as expected. LGTM!

Dominik-Weinzierl commented:
@bassner no :)

willergeorg reviewed (approved):
Tested on TS3. Worked as expected.

nairaarnez reviewed (approved):
Tested in Ts1, everything works as expected

max-bergmann reviewed (approved):
Tested on TS2. Works as expected

Dominik-Weinzierl commented:
@krusche I checked the queries (with `show-sql: true`) and the criteria builder creates the queries as we expect - e.g:

- Java:

  ```java
  String extendedSearchTerm = "%" + searchTerm + "%";
  return (root, query, criteriaBuilder) -> {
      Predicate login = criteriaBuilder.like(root.get(User_.LOGIN), extendedSearchTerm);
      Predicate email = criteriaBuilder.like(root.get(User_.EMAIL), extendedSearchTerm);
      Predicate firstName = criteriaBuilder.like(root.get(User_.FIRST_NAME), extendedSearchTerm);
      Predicate lastName = criteriaBuilder.like(root.get(User_.LAST_NAME), extendedSearchTerm);
  
      return criteriaBuilder.or(login, email, firstName, lastName);
  };
  ```

- Query:
  ```
  select distinct user from jhi_user user
  where user.login like ? or user.email like ? or user.first_name like ? or user.last_name like ? order by user..id 
  ```

TheTrueVester reviewed (dismissed):
Re-tested on ts1. Everything worked as expected.

nadinesamyy reviewed (approved):
tested on ts2, works as described

pantageepapa reviewed (dismissed):
Re-tested on ts2 and it works as expected :)

Dominik-Weinzierl commented:
@krusche I tried to replace both subqueries but I think they are not replaceable, because we need to match all specified values, not just one. Therefore, as far as I know, there is no other solution - but maybe you can help me with more expertise :D 

Subqueries in: `getAllUsersMatchingAuthorities` and `getAllUsersMatchingCourses`

b-fein reviewed (commented):
Code looks good, I only found one issue with a missing translation.

For the queries I have to think about if we could avoid the sub-query if we select the filtered courses and then join the user table onto it and then query for distinct users, instead of joining the other way round and searching for matching courses for every user.
(Should be applicable similarly then also to the authorities query.)

Do you think that might work?

Dominik-Weinzierl commented:
> For the queries I have to think about if we could avoid the sub-query if we select the filtered courses and then join the user table onto it and then query for distinct users, instead of joining the other way round and searching for matching courses for every user.
(Should be applicable similarly then also to the authorities query.)

I guess we have than the problem that we find all users that match one course but not all selected courses :thinking:

Dominik-Weinzierl commented:
@b-fein i guess i found a solution - but i am not sure if this is the right way :D

Dominik-Weinzierl commented:
Edit: Yes this change breaks the combination of multiple filters (e.g. authorities and courses)

Dominik-Weinzierl commented:
Okay - let me explain the fix/workaround and issue. If we filter for authorities (Role) and for courses, we can't use joins for both queries since the group by does not work correctly anymore. 

E.g. we filter for authority Admin and Course A, we want to get all users which match those requirements. Since users can have multiple authorities, we wont get the expected results since the combined group by clause looks like

```sql
GROUP  BY user.id
HAVING Count(groups) = 1
       AND Count(authorities) = 1
```

Therefore i decide to implement a fallback (which use one subquery for authorities) if users filter for courses and authorities. In ALL other cases we use only joins :+1:

b-fein reviewed (approved):
Thanks, the code and the queries look good now. :+1: 

As the queries changed again, though, I think at least one manual (re-)review is needed before the PR can be merged.

krusche commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -177,18 +187,237 @@ public interface UserRepository extends JpaRepository<User, Long> {
+            Predicate firstName = criteriaBuilder.like(root.get(User_.FIRST_NAME), extendedSearchTerm);
+            Predicate lastName = criteriaBuilder.like(root.get(User_.LAST_NAME), extendedSearchTerm);
+
+            return criteriaBuilder.or(login, email, firstName, lastName);
```
This code does not look very nice. There is a lot of repetition. Is it not possible to write this more concise?

krusche replied:
I suggest to move this into a new class UserSpecs as shown in the example on https://stackoverflow.com/a/58243100

Dominik-Weinzierl replied:
Good idea :+1: solved in 5a1cc492bc2ccdbb0239e05f68649d73a6bb7f20 and 0b634275d7aaa550894edb28eb4ff30b7201b8e9

krusche commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -177,18 +187,237 @@ public interface UserRepository extends JpaRepository<User, Long> {
+    }
+
+    /**
+     * Creates the specification to match the provides authorities.
```
this comment is not very helpful. please provide something more descriptive

Dominik-Weinzierl replied:
Yes sure - updated :+1:

Dominik-Weinzierl commented:
I finally found the issue why the join did not work as expected in all cases. Now we have 0 subqueries and as you told me @krusche it is possible :D Sorry for delay.

@b-fein could you please recheck the code changes :) 

@reviewers Since we have changed the queries, we need to test them thoroughly again. :+1: Thank you!

babyygemperor reviewed (approved):
Tested on ts2 again, works as described

b-fein commented on src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:
```diff
@@ -0,0 +1,225 @@
+        String extendedSearchTerm = "%" + searchTerm + "%";
+        return (root, query, criteriaBuilder) -> {
+            String[] columns = { User_.LOGIN, User_.EMAIL, User_.FIRST_NAME, User_.LAST_NAME };
+            Predicate[] predicates = Arrays.stream(columns).map((column) -> criteriaBuilder.like(root.get(column), extendedSearchTerm)).toArray(Predicate[]::new);
```
```suggestion
            Predicate[] predicates = Arrays.stream(columns).map(column -> criteriaBuilder.like(root.get(column), extendedSearchTerm)).toArray(Predicate[]::new);
```
The parentheses are not needed.

b-fein commented on src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:
```diff
@@ -0,0 +1,225 @@
+    private static void updateAllUsersMatchingCoursesJoin(CriteriaBuilder criteriaBuilder, Root<Course> courseRoot, Join<User, String> userToGroupJoin) {
+        // Select all possible group types
+        String[] columns = { Course_.STUDENT_GROUP_NAME, Course_.TEACHING_ASSISTANT_GROUP_NAME, Course_.EDITOR_GROUP_NAME, Course_.INSTRUCTOR_GROUP_NAME };
+        Predicate[] predicates = Arrays.stream(columns).map((column) -> criteriaBuilder.in(courseRoot.get(column)).value(userToGroupJoin)).toArray(Predicate[]::new);
```
```suggestion
        Predicate[] predicates = Arrays.stream(columns).map(column -> criteriaBuilder.in(courseRoot.get(column)).value(userToGroupJoin)).toArray(Predicate[]::new);
```

max-bergmann reviewed (dismissed):
Tested on TS2 again. Add/Remove users works. Filters also work as expected.

b-fein commented on src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:
```diff
@@ -0,0 +1,225 @@
+    static String FILTER_NO_AUTHORITY = "ROLE_NO_AUTHORITY";
+
+    /**
+     * Creates the specification to match the provided search term.
```
```suggestion
     * Creates the specification to match the provided search term within the user’s login, email, and name attributes.
```

b-fein commented on src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * Creates the specification to match the users course.
```
```suggestion
     * Creates the specification to find all users that are part of any of the given courses.
```
Was copy-pasted from the other method.

b-fein commented on src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * Helper method to update the course join. We join the users with the course via the groups of the users and the authority groups of the courses.
```
```suggestion
     * Helper method to update the course join. We join the users with the course via the groups of the users and the authority groups of the courses.
     *
```
Formatting consistency with an empty line between comment and first `@param`

b-fein commented on src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * Creates the specification to match the users course.
```
```suggestion
     * Creates the specification to find users that are part of any of the given courses.
```

jbllhff reviewed (approved):
tested on ts2, works great. When I filter for one user, view the user and go back to the user overview, it is not filtered for the user anymore. Maybe this is desirable in the future?

Dominik-Weinzierl commented:
@b-fein thanks for you suggestions - should be all implemented :+1:

b-fein reviewed (approved):
Thanks, code looks good now without any subqueries. :+1:

Dominik-Weinzierl commented:
Today in the evening i perform an additional small testing session where we check this PR again, afterwards we can mark this PR as ready for merge 👍

rhubinger reviewed (approved):
Tested on TS3. Works as expected.

babyygemperor reviewed (approved):
Tested (again) on ts2 in a testing session, works as described

TheTrueVester reviewed (approved):
Tested on ts2. Everything worked as expected.
